homeassistant/components/transmission/__init__.py

```diff
@@ -55,97 +47,179 @@
+        self.tm_data = None
+        self.unsub_timer = None
+
+    async def async_setup(self):
+        """Set up the Transmission client."""
+
+        config = {
+            CONF_HOST: self.config_entry.data[CONF_HOST],
+            CONF_PORT: self.config_entry.data[CONF_PORT],
+            CONF_USERNAME: self.config_entry.data.get(CONF_USERNAME),
+            CONF_PASSWORD: self.config_entry.data.get(CONF_PASSWORD),
+        }
+        api = await get_api(**config)
+        if not api:
+            return False
+
+        self.tm_data = self.hass.data[DOMAIN][DATA_TRANSMISSION] = TransmissionData(
+            self.hass, self.config_entry, api
+        )
 
-    def add_torrent(service):
-        """Add new torrent to download."""
-        torrent = service.data[ATTR_TORRENT]
-        if torrent.startswith(
-            ("http", "ftp:", "magnet:")
-        ) or hass.config.is_allowed_path(torrent):
-            api.add_torrent(torrent)
-        else:
-            _LOGGER.warning(
-                "Could not add torrent: " "unsupported type or no permission"
+        self.hass.async_add_executor_job(self.tm_data.init_torrent_list)
+        self.hass.async_add_executor_job(self.tm_data.update)
+        self.set_scan_interval(self.scan_interval)
+
+        for platform in ["sensor", "switch"]:
+            self.hass.async_create_task(
+                self.hass.config_entries.async_forward_entry_setup(
+                    self.config_entry, platform
+                )
             )
 
-    hass.services.register(
-        DOMAIN, SERVICE_ADD_TORRENT, add_torrent, schema=SERVICE_ADD_TORRENT_SCHEMA
-    )
+        def add_torrent(service):
+            """Add new torrent to download."""
+            torrent = service.data[ATTR_TORRENT]
+            if torrent.startswith(
+                ("http", "ftp:", "magnet:")
+            ) or self.hass.config.is_allowed_path(torrent):
+                api.add_torrent(torrent)
+            else:
+                _LOGGER.warning(
+                    "Could not add torrent: unsupported type or no permission"
+                )
+
+        self.hass.services.async_register(
+            DOMAIN, SERVICE_ADD_TORRENT, add_torrent, schema=SERVICE_ADD_TORRENT_SCHEMA
+        )
+
+        self.config_entry.add_update_listener(self.async_options_updated)
 
-    sensorconfig = {
-        "sensors": config[DOMAIN][CONF_MONITORED_CONDITIONS],
-        "client_name": config[DOMAIN][CONF_NAME],
-    }
+        return True
 
-    discovery.load_platform(hass, "sensor", DOMAIN, sensorconfig, config)
+    def set_scan_interval(self, scan_interval):
+        """Update scan interval."""
 
-    if config[DOMAIN][TURTLE_MODE]:
-        discovery.load_platform(hass, "switch", DOMAIN, sensorconfig, config)
+        @callback
+        def refresh(event_time):
+            """Get the latest data from Transmission."""
+            self.hass.async_add_executor_job(self.tm_data.update)
```

Comment:
Just call `self.tm_data.update` here. The event helper will make sure its scheduled on the executor thread pool.